### PR TITLE
APPSRE-6941 saas auto promotions manager integration

### PR DESCRIPTION
**Pull request by fishi0x01**
Due to the nature and complexity of the new integration, this PR is humongous and likely requires guided review - I do not dare put this on someone alone.

New integration `saas-auto-promotions-manager`. This integration does:

- fetch publisher state
  - fetch commit shas for refs in publisher saas-files
  - fetch deployment info and target_config hashes from S3
- compute desired subscriber refs and config_hashes based on publisher state
- mr management:
  - remove outdated MRs (we might have newer shas/hashes by now)
  - remove MRs with merge conflicts
  - open a new MR for every new desired subscriber state

We have an enforced `dry-run` for now, so we can evaluate how it behaves in prod w/o doing any changes.

More than half of the files in this PR are for tests.

I want to stress that **this is merely the first iteration** of this integration. We already know that there will be changes and new features added shortly. Further, it is likely we will require a certain amount of refactors to keep this "nice". However, we must start testing this in the real field to get going.

### CODE REVIEW

File: reconcile/saas_auto_promotions_manager/integration.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    deployment_info_cache: dict[str, dict[str, Any]] = {}
+    for channel in saas_channels:
+        has_auto_promotion = False
+        for sub in channel.subscribers:
+            has_auto_promotion |= sub.auto_promote or False
+        if not has_auto_promotion:
+            continue
+        cache_key = (
+            f"{channel.name}/{channel.publisher.repo_url}/{channel.publisher.ref}"
+        )
```
Comment (patjlm) [inline reconcile/saas_auto_promotions_manager/integration.py:123]: could be a tuple?

File: reconcile/saas_auto_promotions_manager/integration.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+            )
+        deployment_info = deployment_info_cache[cache_key]
+        if not deployment_info.get("success"):
+            print(
+                f"[{channel.name}] Commit sha {publisher_commit_sha} wasnt successfully deployed"
+            )
+            continue
+
+        ans.append(
+            PublisherCurrentState(
```
Comment (patjlm) [inline reconcile/saas_auto_promotions_manager/integration.py:154]: I think I would revert this logic to returning a list of `SubscriberCurrentState` to be able to properly handle multiple subscription (APPSRE-7307)

Comment (fishi0x01) [reply reconcile/saas_auto_promotions_manager/integration.py:154]: Makes a lot of sense :+1: I will make it Subscriber centric once I start working on it properly again (once the design doc is merged).

File: reconcile/saas_auto_promotions_manager/integration.py (hunk 1 above)
Comment (fishi0x01) [reply reconcile/saas_auto_promotions_manager/integration.py:125]: Fixed

**Comment by fishi0x01** [comment]
I will have a guided review on this with @chassing tomorrow. Thank you in advance! :bow: :heart:

**Review by chassing — COMMENTED** [review]
In general, please add some `__doc__`s to all important and public class methods.

**Review by chassing — APPROVED** [review]
LGTM. Thanks for all your effort 🙇

**Comment by patjlm** [comment]
wow... that's a lot of code..

**Comment by fishi0x01** [comment]
It is running now in prod and logging promotions it would do. So far it looks very good :)

**Pull request merged**